exemption.php

```diff
@@ -0,0 +1,73 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Page to manage user exemptions.
+ *
+ * @package     factor_exemption
+ * @author      Peter Burnett <peterburnett@catalyst-au.net>
+ * @copyright   Catalyst IT 2023
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+require_once(__DIR__ . '/../../../../../config.php'); require_once($CFG->libdir . '/adminlib.php');
+
+$delete = optional_param('delete', 0, PARAM_INT);
+$extend = optional_param('extend', 0, PARAM_INT);
+
+$context = \context_system::instance();
+$PAGE->set_context($context);
+// We need to load the full admin tree into memory here, otherwise our node isnt locatable as it is loaded conditionally.
+admin_get_root(false, true);
+admin_externalpage_setup('factorexemptionmanageexemptions');
+$url = new moodle_url('/admin/tool/mfa/factor/exemption/exemption.php');
+
+// Process table actions if any.
+if ($extend !== 0) {
+    \factor_exemption\factor::extend_exemption($extend);
+    redirect($url);
+}
+
+if ($delete !== 0) {
+    \factor_exemption\factor::delete_exemption($delete);
+    redirect($url);
+}
+
+$form = new \factor_exemption\form\exemption();
+
+if ($form->is_cancelled()) {
+    redirect('/');
+} else if ($fromform = $form->get_data()) {
+    $user = core_user::get_user_by_username($fromform->user);
+    if (!$user) {
+        $user = core_user::get_user_by_email($fromform->user);
+    }
+    \factor_exemption\factor::add_exemption($user);
```

Comment:
What if $user is still not found here ? I know its checked in the form validation, but there could be for e.g. race conditions where it breaks.

Reply:
Sure thing. Im not too worried, because this will then break on the typeguard for the add_exemption method, but I will rip it out into a static, and wrap it in a type guard and handle it